packages/core/src/lib/utils/session.ts

```diff
@@ -0,0 +1,50 @@
+import { AdapterUser } from "../../adapters";
+import { JWTSessionError, SessionTokenError } from "../../errors";
+import { InternalOptions } from "../../types";
+import { SessionStore } from "./cookie";
+
+// /**
+//  * Returns the currently logged in user, if any.
+//  */
+export async function getLoggedInUser(
+  options: InternalOptions,
+  sessionStore: SessionStore
+): Promise<AdapterUser | null> {
+  const {
+    adapter,
+    jwt,
+    logger,
+    session: { strategy: sessionStrategy },
+  } = options
+
+  const sessionToken = sessionStore.value
+  if (!sessionToken) return null
+
+  // Try to decode JWT
+  if (sessionStrategy === "jwt") {
+    try {
+      const salt = options.cookies.sessionToken.name
+      const payload = await jwt.decode({ ...jwt, token: sessionToken, salt })
+
+      if (!payload || !payload.sub) throw new Error("Invalid JWT")
+
+      return adapter?.getUser(payload.sub) ?? null
+    } catch (e) {
+      logger.error(new JWTSessionError(e as Error))
```

Comment:
We need to destroy the broken session cookie here too
https://github.com/nextauthjs/next-auth/blob/9724c208f3eee09bf625e7fd4cd91f3a80fa28b1/packages/core/src/errors.ts#L218-L224